tests/Functional/LegacyAPITest.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+namespace Doctrine\DBAL\Tests\Functional;
+
+use Doctrine\DBAL\FetchMode;
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\DBAL\Tests\FunctionalTestCase;
+
+use function array_change_key_case;
+use function array_map;
+
+use const CASE_LOWER;
+
+class LegacyAPITest extends FunctionalTestCase
+{
+    /** @var bool */
+    private static $generated = false;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        if (self::$generated !== false) {
+            return;
+        }
+
+        $table = new Table('legacy_table');
+        $table->addColumn('test_int', 'integer');
+        $table->addColumn('test_string', 'string');
+        $table->setPrimaryKey(['test_int']);
+
+        $sm = $this->connection->getSchemaManager();
+        $sm->createTable($table);
+
+        $this->connection->insert('legacy_table', [
+            'test_int' => 1,
+            'test_string' => 'foo',
+        ]);
+
+        self::$generated = true;
+    }
+
+    public function tearDown(): void
+    {
+        $this->connection->executeStatement('DELETE FROM legacy_table WHERE test_int > 1');
+    }
+
+    public function testFetchWithModes(): void
+    {
+        $sql = 'SELECT test_int FROM legacy_table WHERE test_int = 1';
+
+        $stmt = $this->connection->executeQuery($sql);
+        $row  = array_change_key_case($stmt->fetch(FetchMode::ASSOCIATIVE), CASE_LOWER);
+        self::assertEquals(1, $row['test_int']);
+
+        $stmt = $this->connection->executeQuery($sql);
+        $row  = $stmt->fetch(FetchMode::NUMERIC);
+        self::assertEquals(1, $row[0]);
+
+        $stmt = $this->connection->executeQuery($sql);
+        $row  = $stmt->fetch(FetchMode::COLUMN);
+        self::assertEquals(1, $row);
+    }
+
+    public function testFetchAllWithModes(): void
+    {
+        $sql = 'SELECT test_int FROM legacy_table WHERE test_int = 1';
+
+        $stmt = $this->connection->executeQuery($sql);
+
+        $rows = $stmt->fetchAll(FetchMode::ASSOCIATIVE);
+        $rows = array_map(static function ($row) {
+            return array_change_key_case($row, CASE_LOWER);
+        }, $rows);
+
+        self::assertEquals([['test_int' => 1]], $rows);
+
+        $stmt = $this->connection->executeQuery($sql);
+        $rows = $stmt->fetchAll(FetchMode::NUMERIC);
+        self::assertEquals([[0 => 1]], $rows);
+
+        $stmt = $this->connection->executeQuery($sql);
+        $rows = $stmt->fetchAll(FetchMode::COLUMN);
```

Comment:
Should this be a different test or at least a data provider item? If the test for one mode fails, there's no reason to not test the other(s).